lib/puppet/pops/lookup/hiera_config.rb

```diff
@@ -71,25 +57,193 @@ def self.legacy_function_config(function_name)
     )
   end
 
-  def self.config_type
-    @@CONFIG_TYPE ||= create_config_type
-  end
-
   def self.config_exist?(config_root)
     config_path = config_root + CONFIG_FILE_NAME
     config_path.exist?
   end
 
-  def self.create_config_type
+  def self.symkeys_to_string(struct)
+    case(struct)
+    when Hash
+      Hash[struct.map { |k,v| [k.to_s, symkeys_to_string(v)] }]
+    when Array
+      struct.map { |v| symkeys_to_string(v) }
+    else
+      struct
+    end
+  end
+
+  # Creates a new HieraConfig from the given _config_root_. This is where the 'hiera.yaml' is expected to be found
+  # and is also the base location used when resolving relative paths.
+  #
+  # @param config_root [Pathname] Absolute path to the configuration root
+  # @return [LookupConfiguration] the configuration
+  def self.create(config_root)
+    if config_root.is_a?(Hash)
+      config_path = nil
+      loaded_config = config_root
+    else
+      config_path = config_root + CONFIG_FILE_NAME
+      if config_path.exist?
+        loaded_config = YAML.load_file(config_path)
+      else
+        config_path = nil
+        loaded_config = HieraConfigV5::DEFAULT_CONFIG_HASH
+      end
+    end
+
+    version = loaded_config[KEY_VERSION] || loaded_config[:version]
+    version = version.nil? ? 3 : version.to_i
+    case version
+    when 5
+      HieraConfigV5.new(config_root, config_path, loaded_config)
+    when 4
+      HieraConfigV4.new(config_root, config_path, loaded_config)
+    else
+      raise Puppet::DataBinding::LookupError, "#{@config_path}: This runtime does not support #{CONFIG_FILE_NAME} version '#{version}'"
+    end
+  end
+
+  attr_reader :config_path, :version
+
+  # Creates a new HieraConfig from the given _config_root_. This is where the 'lookup.yaml' is expected to be found
+  # and is also the base location used when resolving relative paths.
+  #
+  # @param config_path [Pathname] Absolute path to the configuration
+  # @param loaded_config [Hash] the loaded configuration
+  def initialize(config_root, config_path, loaded_config)
+    @config_root = config_root
+    @config_path = config_path
+    @loaded_config = loaded_config
+    @config = validate_config(self.class.symkeys_to_string(@loaded_config))
+  end
+
+  # Creates the data providers for this config
+  #
+  # @param lookup_invocation [Invocation] Invocation data containing scope, overrides, and defaults
+  # @param parent_data_provider [DataProvider] The data provider that loaded this configuration
+  # @return [Array<DataProvider>] the created providers
+  def create_configured_data_providers(lookup_invocation, parent_data_provider)
+    self.class.not_implemented(self, 'create_configured_data_providers')
+  end
+
+  def validate_config(config)
+    self.class.not_implemented(self, 'validate_config')
+  end
+
+  def version
+    self.class.not_implemented(self, 'version')
+  end
+
+  def name
+    "hiera version #{version}"
```

Comment:
This should probably say 'hiera.yaml version' as opposed to "hiera version" as that may be confused with the hiera gem's version.